registration/registration.go

```diff
@@ -0,0 +1,362 @@
+//   - registering challenges,
+//   - scheduling rounds,
+//   - feeding workers with membership tree roots for proof generation,
+//   - serving proofs.
+type Registration struct {
+	genesis  time.Time
+	cfg      Config
+	roundCfg round_config.Config
+	dbdir    string
+	privKey  ed25519.PrivateKey
+
+	openRoundMutex sync.RWMutex
+	openRound      *round
+
+	db *database
+
+	powVerifiers powVerifiers
+	workerSvc    WorkerService
+}
+
+type newRegistrationOptionFunc func(*newRegistrationOptions)
+
+type newRegistrationOptions struct {
+	powVerifier PowVerifier
+	privKey     ed25519.PrivateKey
+	roundCfg    round_config.Config
+	cfg         Config
+}
+
+func WithPowVerifier(verifier PowVerifier) newRegistrationOptionFunc {
+	return func(opts *newRegistrationOptions) {
+		opts.powVerifier = verifier
+	}
+}
+
+func WithPrivateKey(privKey ed25519.PrivateKey) newRegistrationOptionFunc {
+	return func(opts *newRegistrationOptions) {
+		opts.privKey = privKey
+	}
+}
+
+func WithRoundConfig(roundCfg round_config.Config) newRegistrationOptionFunc {
+	return func(opts *newRegistrationOptions) {
+		opts.roundCfg = roundCfg
+	}
+}
+
+func WithConfig(cfg Config) newRegistrationOptionFunc {
+	return func(opts *newRegistrationOptions) {
+		opts.cfg = cfg
+	}
+}
+
+func NewRegistration(
+	ctx context.Context,
+	genesis time.Time,
+	dbdir string,
+	workerSvc WorkerService,
+	opts ...newRegistrationOptionFunc,
+) (*Registration, error) {
+	options := newRegistrationOptions{
+		roundCfg: round_config.DefaultConfig(),
+		cfg:      DefaultConfig(),
+	}
+	for _, opt := range opts {
+		opt(&options)
+	}
+	if options.privKey == nil {
+		logging.FromContext(ctx).Info("generating new keys")
+		_, priv, err := ed25519.GenerateKey(nil)
+		if err != nil {
+			return nil, fmt.Errorf("generating private key: %w", err)
+		}
+		options.privKey = priv
+	}
+
+	dbPath := filepath.Join(dbdir, "proofs")
+	db, err := newDatabase(dbPath, options.privKey.Public().(ed25519.PublicKey))
+	if err != nil {
+		return nil, fmt.Errorf("opening proofs database: %w", err)
+	}
+
+	r := &Registration{
+		genesis:   genesis,
+		cfg:       options.cfg,
+		roundCfg:  options.roundCfg,
+		dbdir:     dbdir,
+		privKey:   options.privKey,
+		db:        db,
+		workerSvc: workerSvc,
+	}
+
+	if options.powVerifier == nil {
+		r.setupPowProviders(ctx)
+	} else {
+		r.powVerifiers = powVerifiers{current: options.powVerifier}
+	}
+
+	epoch := r.roundCfg.OpenRoundId(r.genesis, time.Now())
+	round, err := newRound(epoch, r.dbdir, r.newRoundOpts()...)
+	if err != nil {
+		return nil, fmt.Errorf("creating new round: %w", err)
+	}
+	logging.FromContext(ctx).Info("opened round", zap.Uint("epoch", epoch), zap.Int("members", round.members))
+	r.openRound = round
+
+	return r, nil
+}
+
+func (r *Registration) Pubkey() ed25519.PublicKey {
+	return r.privKey.Public().(ed25519.PublicKey)
+}
+
+func (r *Registration) Close() (res error) {
+	return errors.Join(res, r.db.Close(), r.openRound.Close())
+}
```

Comment:
```suggestion
func (r *Registration) Close() error {
	return errors.Join(r.db.Close(), r.openRound.Close())
}
```

Reply:
What's the benefit?

Reply:
True, I'm dumb and somehow looked at the suggested diff in the inverted direction 😅